src/listener.ts

```diff
@@ -8,100 +8,268 @@ import { writeFromReadableStream } from './utils'
+  value: newResponse,
+})
 
-    const init = {
-      method: method,
-      headers: headerRecord,
-    } as RequestInit
+function newRequestFromIncoming(
+  method: string,
+  url: string,
+  incoming: IncomingMessage | Http2ServerRequest
+): Request {
+  const headerRecord: [string, string][] = []
+  const len = incoming.rawHeaders.length
+  for (let i = 0; i < len; i += 2) {
+    headerRecord.push([incoming.rawHeaders[i], incoming.rawHeaders[i + 1]])
+  }
 
-    if (!(method === 'GET' || method === 'HEAD')) {
-      // lazy-consume request body
-      init.body = Readable.toWeb(incoming) as ReadableStream<Uint8Array>
-      // node 18 fetch needs half duplex mode when request body is stream
-      ;(init as any).duplex = 'half'
-    }
+  const init = {
+    method: method,
+    headers: headerRecord,
+  } as RequestInit
+
+  if (!(method === 'GET' || method === 'HEAD')) {
+    // lazy-consume request body
+    init.body = Readable.toWeb(incoming) as ReadableStream<Uint8Array>
+    // node 18 fetch needs half duplex mode when request body is stream
+    ;(init as any).duplex = 'half'
+  }
+
+  return new Request(url, init)
+}
+
+const requestPrototype: Record<string, any> = {
+  getRequestCache() {
+    return (this.requestCache ||= newRequestFromIncoming(this.method, this.url, this.incoming))
+  },
+  get body() {
+    return this.getRequestCache().body
+  },
+  get bodyUsed() {
+    return this.getRequestCache().bodyUsed
+  },
+  get cache() {
+    return this.getRequestCache().cache
+  },
+  get credentials() {
+    return this.getRequestCache().credentials
+  },
+  get destination() {
+    return this.getRequestCache().destination
+  },
+  get headers() {
+    return this.getRequestCache().headers
+  },
+  get integrity() {
+    return this.getRequestCache().integrity
+  },
+  get mode() {
+    return this.getRequestCache().mode
+  },
+  get redirect() {
+    return this.getRequestCache().redirect
+  },
+  get referrer() {
+    return this.getRequestCache().referrer
+  },
+  get referrerPolicy() {
+    return this.getRequestCache().referrerPolicy
+  },
+  get signal() {
+    return this.getRequestCache().signal
+  },
+  arrayBuffer() {
+    return this.getRequestCache().arrayBuffer()
+  },
+  blob() {
+    return this.getRequestCache().blob()
+  },
+  clone() {
+    return this.getRequestCache().clone()
+  },
+  formData() {
+    return this.getRequestCache().formData()
+  },
+  json() {
+    return this.getRequestCache().json()
+  },
+  text() {
+    return this.getRequestCache().text()
+  },
+}
+
+const handleError = (e: unknown): Response =>
+  new Response(null, {
+    status:
+      e instanceof Error && (e.name === 'TimeoutError' || e.constructor.name === 'TimeoutError')
+        ? 504 // timeout error emits 504 timeout
+        : 500,
+  })
+
+const responseViaResponseObject = async (
+  res: Response | Promise<Response>,
+  outgoing: ServerResponse | Http2ServerResponse
+) => {
+  try {
+    res = await res
```

Comment:
Can we make it does not `await` if it's not necessary?

Reply:
@usualoma Ah, you already did it!
https://github.com/usualoma/node-server/pull/2/files